File: src/main/java/org/moinex/ui/dialog/investment/BaseCryptoExchangeManagement.java
/*
 * Filename: BaseCryptoExchangeManagement.java
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import java.math.BigDecimal;
import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.DatePicker;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.investment.Ticker;
import org.moinex.service.CalculatorService;
import org.moinex.service.TickerService;
import org.moinex.ui.common.CalculatorController;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.moinex.util.enums.TickerType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;

/**
 * Base class to implement the common behavior of the Add and Edit Crypto Exchange
 */
@NoArgsConstructor
public abstract class BaseCryptoExchangeManagement {
    @FXML protected Label cryptoSoldAfterBalanceValueLabel;

    @FXML protected Label cryptoReceivedAfterBalanceValueLabel;

    @FXML protected Label cryptoSoldCurrentBalanceValueLabel;

    @FXML protected Label cryptoReceivedCurrentBalanceValueLabel;

    @FXML protected ComboBox<Ticker> cryptoSoldComboBox;

    @FXML protected ComboBox<Ticker> cryptoReceivedComboBox;

    @FXML protected TextField descriptionField;

    @FXML protected TextField cryptoSoldQuantityField;

    @FXML protected TextField cryptoReceivedQuantityField;

    @FXML protected DatePicker exchangeDatePicker;

    protected ConfigurableApplicationContext springContext;

    protected TickerService tickerService;

    protected CalculatorService calculatorService;

    protected List<Ticker> cryptos;

    protected Ticker fromCrypto = null;

    /**
     * Constructor
     * @param tickerService TickerService
     * @param calculatorService CalculatorService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    protected BaseCryptoExchangeManagement(
            TickerService tickerService, CalculatorService calculatorService) {
        this.tickerService = tickerService;
        this.calculatorService = calculatorService;
    }

    public void setFromCryptoComboBox(Ticker tk) {
        if (cryptos.stream().noneMatch(c -> c.getId().equals(tk.getId()))) {
            return;
        }

        this.fromCrypto = tk;
        cryptoSoldComboBox.setValue(fromCrypto);
        updateFromCryptoCurrentQuantity();
    }

    @FXML
    protected void initialize() {
        configureComboBoxes();

        loadCryptosFromDatabase();
        populateCryptoComboBoxes();

        // Configure the date picker
        UIUtils.setDatePickerFormat(exchangeDatePicker);

        // Reset all labels
        UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel);
        UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel);
        UIUtils.resetLabel(cryptoSoldCurrentBalanceValueLabel);
        UIUtils.resetLabel(cryptoReceivedCurrentBalanceValueLabel);

        cryptoSoldComboBox.setOnAction(
                e -> {
                    updateFromCryptoCurrentQuantity();
                    updateFromCryptoQuantityAfterExchange();
                });

        cryptoReceivedComboBox.setOnAction(
                e -> {
                    updateToCryptoCurrentQuantity();
                    updateToCryptoQuantityAfterExchange();
                });

        configureListeners();
    }

    @FXML
    protected abstract void handleSave();

    @FXML
    protected void handleCancel() {
        Stage stage = (Stage) cryptoReceivedQuantityField.getScene().getWindow();
        stage.close();
    }

    @FXML
    protected void handleCryptoSoldOpenCalculator() {
        WindowUtils.openPopupWindow(
                Constants.CALCULATOR_FXML,
                "Calculator",
                springContext,
                (CalculatorController controller) -> {},
                List.of(
                        () ->
                                calculatorService.updateComponentWithResult(
                                        cryptoSoldQuantityField)));
    }

    @FXML
    protected void handleCryptoReceivedOpenCalculator() {
        WindowUtils.openPopupWindow(
                Constants.CALCULATOR_FXML,
                "Calculator",
                springContext,
                (CalculatorController controller) -> {},
                List.of(
                        () ->
                                calculatorService.updateComponentWithResult(
                                        cryptoReceivedQuantityField)));
    }

    protected void updateFromCryptoCurrentQuantity() {
        Ticker cryptoSold = cryptoSoldComboBox.getValue();

        if (cryptoSold == null) {
            return;
        }

        if (cryptoSold.getCurrentQuantity().compareTo(BigDecimal.ZERO) < 0) {
            // Must be unreachable
            UIUtils.setLabelStyle(
                    cryptoSoldCurrentBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
        } else {
            UIUtils.setLabelStyle(
                    cryptoSoldCurrentBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
        }

        cryptoSoldCurrentBalanceValueLabel.setText(cryptoSold.getCurrentQuantity().toString());
    }

    protected void updateToCryptoCurrentQuantity() {
        Ticker cryptoReceived = cryptoReceivedComboBox.getValue();

        if (cryptoReceived == null) {
            return;
        }

        if (cryptoReceived.getCurrentQuantity().compareTo(BigDecimal.ZERO) < 0) {
            // Must be unreachable
            UIUtils.setLabelStyle(
                    cryptoReceivedCurrentBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
        } else {
            UIUtils.setLabelStyle(
                    cryptoReceivedCurrentBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
        }

        cryptoReceivedCurrentBalanceValueLabel.setText(
                cryptoReceived.getCurrentQuantity().toString());
    }

    protected void updateFromCryptoQuantityAfterExchange() {
        String cryptoSoldQuantityStr = cryptoSoldQuantityField.getText();
        Ticker cryptoSold = cryptoSoldComboBox.getValue();

        if (cryptoSoldQuantityStr == null
                || cryptoSoldQuantityStr.isBlank()
                || cryptoSold == null) {
            UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel);
            return;
        }

        try {
            BigDecimal exchangeQuantity = new BigDecimal(cryptoSoldQuantityStr);

            if (exchangeQuantity.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel);
                return;
            }

            BigDecimal cryptoSoldAfterBalance =
                    cryptoSold.getCurrentQuantity().subtract(exchangeQuantity);

            if (cryptoSoldAfterBalance.compareTo(BigDecimal.ZERO) < 0) {
                // Remove old style and add negative style
                UIUtils.setLabelStyle(
                        cryptoSoldAfterBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                // Remove old style and add neutral style
                UIUtils.setLabelStyle(
                        cryptoSoldAfterBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
            }

            cryptoSoldAfterBalanceValueLabel.setText(cryptoSoldAfterBalance.toString());
        } catch (NumberFormatException e) {
            UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel);
        }
    }

    protected void updateToCryptoQuantityAfterExchange() {
        String cryptoReceivedQuantityStr = cryptoReceivedQuantityField.getText();
        Ticker cryptoReceived = cryptoReceivedComboBox.getValue();

        if (cryptoReceivedQuantityStr == null
                || cryptoReceivedQuantityStr.isBlank()
                || cryptoReceived == null) {
            UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel);
            return;
        }

        try {
            BigDecimal exchangeQuantity = new BigDecimal(cryptoReceivedQuantityStr);

            if (exchangeQuantity.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel);
                return;
            }

            BigDecimal cryptoReceivedAfterBalance =
                    cryptoReceived.getCurrentQuantity().add(exchangeQuantity);

            if (cryptoReceivedAfterBalance.compareTo(BigDecimal.ZERO) < 0) {
                // Remove old style and add negative style
                UIUtils.setLabelStyle(
                        cryptoReceivedAfterBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                // Remove old style and add neutral style
                UIUtils.setLabelStyle(
                        cryptoReceivedAfterBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
            }

            cryptoReceivedAfterBalanceValueLabel.setText(cryptoReceivedAfterBalance.toString());
        } catch (NumberFormatException e) {
            UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel);
        }
    }

    protected void loadCryptosFromDatabase() {
        cryptos = tickerService.getAllNonArchivedTickersByType(TickerType.CRYPTOCURRENCY);
    }

    protected void populateCryptoComboBoxes() {
        cryptoSoldComboBox.getItems().setAll(cryptos);
        cryptoReceivedComboBox.getItems().setAll(cryptos);
    }

    protected void configureComboBoxes() {
        UIUtils.configureComboBox(cryptoSoldComboBox, Ticker::getName);
        UIUtils.configureComboBox(cryptoReceivedComboBox, Ticker::getName);
    }

    protected void configureListeners() {
        // Ensure that the user can only input numbers in the quantity field
        cryptoSoldQuantityField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.INVESTMENT_VALUE_REGEX)) {
                                cryptoSoldQuantityField.setText(oldValue);
                            } else {
                                updateFromCryptoQuantityAfterExchange();
                                updateToCryptoQuantityAfterExchange();
                            }
                        });

        cryptoReceivedQuantityField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.INVESTMENT_VALUE_REGEX)) {
                                cryptoReceivedQuantityField.setText(oldValue);
                            } else {
                                updateFromCryptoQuantityAfterExchange();
                                updateToCryptoQuantityAfterExchange();
                            }
                        });
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/base/BaseCryptoExchangeManagement.kt
/*
 * Filename: BaseCryptoExchangeManagement.kt (original filename: BaseCryptoExchangeManagement.java)
 * Created on: March  9, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.base

import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.DatePicker
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.constant.Files
import org.moinex.common.constant.Styles
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.enums.AssetType
import org.moinex.model.investment.Ticker
import org.moinex.service.CalculatorService
import org.moinex.service.PreferencesService
import org.moinex.service.investment.TickerService
import org.moinex.ui.common.CalculatorController
import org.springframework.context.ConfigurableApplicationContext
import java.math.BigDecimal

abstract class BaseCryptoExchangeManagement(
    protected val tickerService: TickerService,
    protected val calculatorService: CalculatorService,
    protected val preferencesService: PreferencesService,
) {
    @FXML
    protected lateinit var cryptoSoldAfterBalanceValueLabel: Label

    @FXML
    protected lateinit var cryptoReceivedAfterBalanceValueLabel: Label

    @FXML
    protected lateinit var cryptoSoldCurrentBalanceValueLabel: Label

    @FXML
    protected lateinit var cryptoReceivedCurrentBalanceValueLabel: Label

    @FXML
    protected lateinit var cryptoSoldComboBox: ComboBox<Ticker>

    @FXML
    protected lateinit var cryptoReceivedComboBox: ComboBox<Ticker>

    @FXML
    protected lateinit var descriptionField: TextField

    @FXML
    protected lateinit var cryptoSoldQuantityField: TextField

    @FXML
    protected lateinit var cryptoReceivedQuantityField: TextField

    @FXML
    protected lateinit var exchangeDatePicker: DatePicker

    protected lateinit var springContext: ConfigurableApplicationContext
    protected var cryptos: List<Ticker> = emptyList()
    protected lateinit var fromCrypto: Ticker

    fun setFromCryptoComboBox(tk: Ticker) {
        if (cryptos.none { it.id!! == tk.id!! }) {
            return
        }

        fromCrypto = tk
        cryptoSoldComboBox.value = fromCrypto
        updateFromCryptoCurrentQuantity()
    }

    @FXML
    protected open fun initialize() {
        configureComboBoxes()

        loadCryptosFromDatabase()
        populateCryptoComboBoxes()

        UIUtils.setDatePickerFormat(exchangeDatePicker)

        UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel)
        UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel)
        UIUtils.resetLabel(cryptoSoldCurrentBalanceValueLabel)
        UIUtils.resetLabel(cryptoReceivedCurrentBalanceValueLabel)

        cryptoSoldComboBox.setOnAction {
            updateFromCryptoCurrentQuantity()
            updateFromCryptoQuantityAfterExchange()
        }

        cryptoReceivedComboBox.setOnAction {
            updateToCryptoCurrentQuantity()
            updateToCryptoQuantityAfterExchange()
        }

        configureListeners()
    }

    @FXML
    protected abstract fun handleSave()

    @FXML
    protected open fun handleCancel() {
        (cryptoReceivedQuantityField.scene.window as Stage).close()
    }

    @FXML
    protected fun handleCryptoSoldOpenCalculator() {
        WindowUtils.openPopupWindow(
            Files.CALCULATOR_FXML,
            preferencesService.translate(TranslationKeys.INVESTMENT_LABEL_CALCULATOR),
            springContext,
            { _: CalculatorController -> },
            listOf(Runnable { calculatorService.updateComponentWithResult(cryptoSoldQuantityField) }),
        )
    }

    @FXML
    protected fun handleCryptoReceivedOpenCalculator() {
        WindowUtils.openPopupWindow(
            Files.CALCULATOR_FXML,
            preferencesService.translate(TranslationKeys.INVESTMENT_LABEL_CALCULATOR),
            springContext,
            { _: CalculatorController -> },
            listOf(Runnable { calculatorService.updateComponentWithResult(cryptoReceivedQuantityField) }),
        )
    }

    protected fun updateFromCryptoCurrentQuantity() {
        val cryptoSold = cryptoSoldComboBox.value ?: return

        if (cryptoSold.currentQuantity < BigDecimal.ZERO) {
            UIUtils.setLabelStyle(cryptoSoldCurrentBalanceValueLabel, Styles.NEGATIVE_BALANCE_STYLE)
        } else {
            UIUtils.setLabelStyle(cryptoSoldCurrentBalanceValueLabel, Styles.NEUTRAL_BALANCE_STYLE)
        }

        cryptoSoldCurrentBalanceValueLabel.text = cryptoSold.currentQuantity.toString()
    }

    protected fun updateToCryptoCurrentQuantity() {
        val cryptoReceived = cryptoReceivedComboBox.value ?: return

        if (cryptoReceived.currentQuantity < BigDecimal.ZERO) {
            UIUtils.setLabelStyle(cryptoReceivedCurrentBalanceValueLabel, Styles.NEGATIVE_BALANCE_STYLE)
        } else {
            UIUtils.setLabelStyle(cryptoReceivedCurrentBalanceValueLabel, Styles.NEUTRAL_BALANCE_STYLE)
        }

        cryptoReceivedCurrentBalanceValueLabel.text = cryptoReceived.currentQuantity.toString()
    }

    protected fun updateFromCryptoQuantityAfterExchange() {
        val cryptoSoldQuantityStr = cryptoSoldQuantityField.text
        val cryptoSold = cryptoSoldComboBox.value

        if (cryptoSoldQuantityStr.isNullOrBlank() || cryptoSold == null) {
            UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel)
            return
        }

        runCatching {
            val exchangeQuantity = cryptoSoldQuantityStr.toBigDecimal()

            if (exchangeQuantity < BigDecimal.ZERO) {
                UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel)
                return
            }

            val cryptoSoldAfterBalance = cryptoSold.currentQuantity.subtract(exchangeQuantity)

            if (cryptoSoldAfterBalance < BigDecimal.ZERO) {
                UIUtils.setLabelStyle(cryptoSoldAfterBalanceValueLabel, Styles.NEGATIVE_BALANCE_STYLE)
            } else {
                UIUtils.setLabelStyle(cryptoSoldAfterBalanceValueLabel, Styles.NEUTRAL_BALANCE_STYLE)
            }

            cryptoSoldAfterBalanceValueLabel.text = cryptoSoldAfterBalance.toString()
        }.onFailure {
            UIUtils.resetLabel(cryptoSoldAfterBalanceValueLabel)
        }
    }

    protected fun updateToCryptoQuantityAfterExchange() {
        val cryptoReceivedQuantityStr = cryptoReceivedQuantityField.text
        val cryptoReceived = cryptoReceivedComboBox.value

        if (cryptoReceivedQuantityStr.isNullOrBlank() || cryptoReceived == null) {
            UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel)
            return
        }

        runCatching {
            val exchangeQuantity = cryptoReceivedQuantityStr.toBigDecimal()

            if (exchangeQuantity < BigDecimal.ZERO) {
                UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel)
                return
            }

            val cryptoReceivedAfterBalance = cryptoReceived.currentQuantity.add(exchangeQuantity)

            if (cryptoReceivedAfterBalance < BigDecimal.ZERO) {
                UIUtils.setLabelStyle(cryptoReceivedAfterBalanceValueLabel, Styles.NEGATIVE_BALANCE_STYLE)
            } else {
                UIUtils.setLabelStyle(cryptoReceivedAfterBalanceValueLabel, Styles.NEUTRAL_BALANCE_STYLE)
            }

            cryptoReceivedAfterBalanceValueLabel.text = cryptoReceivedAfterBalance.toString()
        }.onFailure {
            UIUtils.resetLabel(cryptoReceivedAfterBalanceValueLabel)
        }
    }

    protected fun loadCryptosFromDatabase() {
        cryptos = tickerService.getAllNonArchivedTickersByType(AssetType.CRYPTOCURRENCY)
    }

    protected fun populateCryptoComboBoxes() {
        cryptoSoldComboBox.items.setAll(cryptos)
        cryptoReceivedComboBox.items.setAll(cryptos)
    }

    protected fun configureComboBoxes() {
        UIUtils.configureComboBox(cryptoSoldComboBox, Ticker::name)
        UIUtils.configureComboBox(cryptoReceivedComboBox, Ticker::name)
    }

    protected fun configureListeners() {
        cryptoSoldQuantityField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.INVESTMENT_VALUE_REGEX))) {
                cryptoSoldQuantityField.text = oldValue
            } else {
                updateFromCryptoQuantityAfterExchange()
                updateToCryptoQuantityAfterExchange()
            }
        }

        cryptoReceivedQuantityField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.INVESTMENT_VALUE_REGEX))) {
                cryptoReceivedQuantityField.text = oldValue
            } else {
                updateFromCryptoQuantityAfterExchange()
                updateToCryptoQuantityAfterExchange()
            }
        }
    }
}
